src/t_hash.c

```diff
@@ -1134,6 +1640,486 @@ void hrandfieldCommand(client *c) {
+    int flag = 0;
+    robj *hashObj, *keyArg = c->argv[1], *expireArg = c->argv[2];
+
+    /* Read the hash object */
+    if ((hashObj = lookupKeyWriteOrReply(c, keyArg, shared.null[c->resp])) == NULL ||
+        checkType(c, hashObj, OBJ_HASH)) return;
+
+    /* not supported yet listpack */
+    if (hashObj->encoding == OBJ_ENCODING_LISTPACK) {
+        addReplyError(c,"Hash field expire for listpack not supported yet.");
+        return;
+    }
+
+    dict *d = hashObj->ptr;
+
+    /* Read the expiry time from command */
+    if (getLongLongFromObjectOrReply(c, expireArg, &expire, NULL) != C_OK)
+        return;
+
+    /* Check expire overflow */
+    if (expire > (long long) EB_EXPIRE_TIME_MAX) {
+        addReplyErrorExpireTime(c);
+        return;
+    }
+
+    if (unit == UNIT_SECONDS) {
+        if (expire > (long long) EB_EXPIRE_TIME_MAX / 1000) {
+            addReplyErrorExpireTime(c);
+            return;
+        }
+        expire *= 1000;
+    } else {
+        if (expire > (long long) EB_EXPIRE_TIME_MAX) {
+            addReplyErrorExpireTime(c);
+            return;
+        }
+    }
+
+    if (expire > (long long) EB_EXPIRE_TIME_MAX - basetime) {
+        addReplyErrorExpireTime(c);
+        return;
+    }
+    expire += basetime;
+
+    /* Read optional flag [NX|XX|GT|LT] */
+    char *optArg = c->argv[3]->ptr;
+    if (!strcasecmp(optArg, "nx")) {
+        flag = HFE_NX; ++numFieldsAt;
+    } else if (!strcasecmp(optArg, "xx")) {
+        flag = HFE_XX; ++numFieldsAt;
+    } else if (!strcasecmp(optArg, "gt")) {
+        flag = HFE_GT; ++numFieldsAt;
+    } else if (!strcasecmp(optArg, "lt")) {
+        flag = HFE_LT; ++numFieldsAt;
+    }
+
+    /* Read number of fields */
+    if (getRangeLongFromObjectOrReply(c, c->argv[numFieldsAt], 1, LONG_MAX,
+                                      &numFields, "Parameter `numFields` should be greater than 0") != C_OK)
+        return;
+
+    /* Verify `numFields` is consistent with number of arguments */
+    if (numFields > (c->argc - numFieldsAt - 1)) {
+        addReplyError(c, "Parameter `numFileds` is more than number of arguments");
+        return;
+    }
+
+    dictExpireMetadata *dictExpireMeta;
+    uint64_t minExpire = EB_EXPIRE_TIME_INVALID;
+
+    /* If dict doesn't have metadata attached */
+    if (!isDictWithMetaHFE(d)) {
+        /* Realloc (only header of dict) with metadata for hash-field expiration */
+        dictTypeAddMeta(&d, &mstrHashDictTypeWithHFE);
+        dictExpireMeta = (dictExpireMetadata *) dictMetadata(d);
+        hashObj->ptr = d;
+
+        /* Find the key in the keyspace. Need to keep reference to the key for
+         * notifications or even removal of the hash */
+        dictEntry *de = dbFind(c->db, keyArg->ptr);
+        serverAssert(de != NULL);
+        sds key = dictGetKey(de);
+
+        /* Fillup dict HFE metadata */
+        dictExpireMeta->key = key;            /* reference key in keyspace */
```

Comment:
Hi @sundb, 
This is great input! 

I totally discarded defrag so far, planning to address it at a later time.

Do you know if there is an issue regarding hash-fields since now they embed linked list?